### PR TITLE
[完成翻译] src/null-safety/migration-guide.md

### LINKED ISSUE #242: Null safety migration guide - 空安全迁移指南
Markdown 文档：https://github.com/cfug/dart.cn/blob/master/src/null-safety/migration-guide.md

文档预览地址：https://dart.cn/null-safety/migration-guide

建议完成天数：5-7 天

Synced from: #240

### PR DESCRIPTION
**Pull request by AlexV525**
Fixes #242 .

### CODE REVIEW

File: src/null-safety/migration-guide.md [hunk 1]
```diff
@@ -243,6 +383,9 @@ The default migration when this code is outside a function
 (it's different within a function)
 is backward compatible but not ideal:
 
+当这些代码处在函数外时，默认的迁移改动是向下兼容的，但并不理想
```
Comment (MeandNi) [inline src/null-safety/migration-guide.md:386]: 与下文 “影响到直接或间接使用了 `zero` 的代码。” 对应，应该是 向后 兼容
```suggestion
当这些代码处在函数外时，默认的迁移改动是向后兼容的，但并不理想
```

File: src/null-safety/migration-guide.md [hunk 2]
```diff
@@ -400,28 +608,44 @@ To migrate a package by hand, follow these steps:
    every library in the package to 2.12,
    opting them all in to null safety.
 
+   在版本低于 `2.12.0-0` 的 SDK 上运行 `dart pub get` 时，
```
Comment (MeandNi) [inline src/null-safety/migration-guide.md:611]: 这部分原意应该是 “SDK 最低版本的限制是 2.12.0-0”，所以使用最低。
```suggestion
   在版本最低是 `2.12.0-0` 的 SDK 上运行 `dart pub get` 时，
```

File: src/null-safety/migration-guide.md [hunk 3]
```diff
@@ -432,36 +656,59 @@ $ dart analyze     # or `flutter analyze`
...
 
+如果您的代码通过了分析，接下来可以开始测试：
+
 ```terminal
 $ dart test       # or `flutter test`
 ```
 
 You might need to update tests that expect null values.
 
+您可能需要更新期望空值的测试。
```
Comment (MeandNi) [inline src/null-safety/migration-guide.md:671]:
```suggestion
您可能需要更新使用了空值作为预期用例的测试代码。
```

File: src/null-safety/migration-guide.md [hunk 4]
```diff
@@ -432,36 +656,59 @@ $ dart analyze     # or `flutter analyze`
...
 
+### SDK 限制
+
 Set the lower SDK constraint to the beta version of 2.12
 that you used to test the migration,
 and the upper SDK constraint to `<3.0.0`.
 For example, if you're using 2.12.0-29.10.beta,
 then your constraints should look like this:
 
+将 SDK 限制的下界设置为您用于测试的 2.12 beta 版本，
```
Comment (MeandNi) [inline src/null-safety/migration-guide.md:708]:
```suggestion
将 SDK 限制的下界设置为您测试迁移时的 2.12 beta 版本，
```

File: src/null-safety/migration-guide.md [hunk 5]
```diff
@@ -471,38 +718,64 @@ With these constraints,
 packages that are published during null safety beta
 can still work with the next stable release of the Dart SDK.
 
+在空安全测试期间以这种形式进行发布的包，能在下一个 Dart SDK 稳定版中继续使用。
 
 ### Package version
 
+### 软件包的保本
```
Comment (MeandNi) [inline src/null-safety/migration-guide.md:725]:
```suggestion
### 软件包的版本
```

File: src/null-safety/migration-guide.md [hunk 6]
```diff
@@ -471,38 +718,64 @@ With these constraints,
...
 ## Welcome to null safety
 
+## 欢迎使用空安全
+
 If you made it this far,
 you should have a fully migrated, null-safe Dart package.
 If all of the packages you depend on are migrated too,
 then your program is sound with respect to null-reference errors.
 
+如果您走到了这一步，您应该拥有了一个完全迁移的空安全 Dart 软件包。
```
Comment (MeandNi) [inline src/null-safety/migration-guide.md:776]:
```suggestion
如果您走到了这一步，您应该已经完全迁移了您的 Dart 软件包至空安全了。
```

File: src/null-safety/migration-guide.md [hunk 7]
```diff
@@ -471,38 +718,64 @@ With these constraints,
...
+当您的所有依赖都也已完成了迁移，那么您的软件就是健全的，同时可以正确处理空引用的错误。
```
Comment (MeandNi) [inline src/null-safety/migration-guide.md:777]:
```suggestion
当您的所有依赖也都这样完成了迁移，那么您的程序就是健全的，同时可以正确处理空引用的错误。
```

File: src/null-safety/migration-guide.md [hunk 8]
```diff
@@ -60,38 +103,58 @@ with the help of the `dart pub outdated` command in null-safety mode.
...
+
 [Dart SDK archive]: /tools/sdk/archive#beta-channel
 
 {{ site.alert.warn }}
+
   If you return to working on production code, remember to
   **switch back to a stable release**
   (for example, by running `flutter channel stable`).
+
+  如果您需要开发别的生产环境的项目，请记得 **将 SDK 切换至稳定版本**
```
Comment (MeandNi) [inline src/null-safety/migration-guide.md:143]:
```suggestion
  如果您想继续投入开发生产环境中的代码，请记得 **将 SDK 切换至稳定版本**
```

File: src/null-safety/migration-guide.md [hunk 9]
```diff
@@ -101,42 +164,72 @@ then you can start migrating.
...
 1. Update `pubspec.yaml` to use null-safe releases
    (as listed in the **Resolvable** column)
    of its dependencies.
    Omit `.x` suffixes to make version solving more flexible,
    and _don't_ update the SDK minimum constraint.
    For example, the `pubspec.yaml` file might look like this:
+
+   更新 `pubspec.yaml`，使用依赖们的空安全版本
+   （在 **Resolvable** 列中列出的）。
+   省略 `.x` 的后缀版本，让版本被更灵活地解析，
```
Comment (MeandNi) [inline src/null-safety/migration-guide.md:229]:
```suggestion
   省略后缀版本号  `.x`，使之能够被更灵活地解析，
```

File: src/null-safety/migration-guide.md [hunk 10]
```diff
@@ -101,42 +164,72 @@ then you can start migrating.
...
+   并且 **不要** 更新 SDK 最低版本显示。
```
Comment (MeandNi) [inline src/null-safety/migration-guide.md:230]:
```suggestion
   并且 **不要** 更新 SDK 最低版本限制。
```

File: src/null-safety/migration-guide.md [hunk 11]
```diff
@@ -152,25 +245,43 @@ update its dependencies to null-safe versions:
...
 * [Migrate your code by hand.](#migrating-by-hand)
 
+  [自己动手，丰衣足食。](#migrating-by-hand)
+
 {{ site.alert.tip }}
+
   For additional help while migrating code, check the
   [null safety FAQ][].
+
+  若您需要迁移过程的额外帮助，请查看 [空安全常见问题][null safety FAQ]。
```
Comment (MeandNi) [inline src/null-safety/migration-guide.md:283]:
```suggestion
  若您在迁移过程中需要额外的帮助，请查看 [空安全常见问题][null safety FAQ]。
```

File: src/null-safety/migration-guide.md [hunk 12]
```diff
@@ -256,43 +399,63 @@ adding the `!`.
 Because you know that `zero` can't be null,
 you can improve the migration result.
 
+点击 **line 3** 链接，您可以看到迁移工具添加 `!` 的原因。
+而因为您只带 `zero` 不会为空，所以您可以改进迁移结果。
```
Comment (MeandNi) [inline src/null-safety/migration-guide.md:403]:
```suggestion
而因为您知道 `zero` 不会为空，所以您可以改进迁移结果。
```

File: src/null-safety/migration-guide.md [hunk 13]
```diff
@@ -256,43 +399,63 @@ adding the `!`.
...
+  在迁移工具的 **Edit Details** 窗格中，您可以通过 **Add `/*?*/` hint** 和
+  **Add `/*!*/` hint** 按钮来添加提示标记。
+
   These buttons add comments to your file immediately,
   and there's **no Undo**.
   If you don't want a hint that the tool inserted,
   you can use your usual code editor to remove it.
 
+  按下这些按钮，相应的标记会立刻添加到代码中，并且 **无法撤销**。
+  如果您想删除标记，可以使用平常使用的代码编辑器删除它。
```
Comment (MeandNi) [inline src/null-safety/migration-guide.md:427]:
```suggestion
  如果您想删除标记，可以和平常一样使用代码编辑器删除它。
```

File: src/null-safety/migration-guide.md (hunk 2 above)
Comment (AlexV525) [reply src/null-safety/migration-guide.md:611]: 翻译时理解错了

**Comment by MeandNi** [comment]
LGTM
